File: src/content/docs/topics/cs/coding-problems/graphs/269-alien-dictionary.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function alienOrder(words: string[]): string {
    // TODO: implement
    return '';
}

function _runTests(): void {
    const r1 = alienOrder(['wrt', 'wrf', 'er', 'ett', 'rftt']);
    assert(r1 === 'wertf', `got ${JSON.stringify(r1)}`);
    const r2 = alienOrder(['z', 'x']);
    assert(r2 === 'zx', `got ${JSON.stringify(r2)}`);
    assert(alienOrder(['z', 'x', 'z']) === '');
    assert(alienOrder(['abc', 'ab']) === '');
    const r3 = alienOrder(['abc']);
    assert(new Set(r3).size === 3 && r3.includes('a') && r3.includes('b') && r3.includes('c'), `got ${JSON.stringify(r3)}`);
    // perf
    const bigWords = Array.from({ length: 200 }, (_, i) =>
        String.fromCharCode(97 + (i % 26)).repeat(3) + String.fromCharCode(97 + ((i + 1) % 26)));
    const t0 = performance.now();
    alienOrder(bigWords);
    console.log(`perf alien-dictionary 200 words: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();

File: src/content/docs/topics/cs/coding-problems/graphs/269-alien-dictionary.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func alienOrder(words []string) string {
	// TODO: implement
	return ""
}

func containsAll(s, chars string) bool {
	for _, c := range chars {
		found := false
		for _, sc := range s {
			if sc == c {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

func runTests() {
	r1 := alienOrder([]string{"wrt", "wrf", "er", "ett", "rftt"})
	assert(r1 == "wertf", fmt.Sprintf("got %q", r1))
	r2 := alienOrder([]string{"z", "x"})
	assert(r2 == "zx", fmt.Sprintf("got %q", r2))
	assert(alienOrder([]string{"z", "x", "z"}) == "")
	assert(alienOrder([]string{"abc", "ab"}) == "")
	r3 := alienOrder([]string{"abc"})
	assert(containsAll(r3, "abc") && len(r3) == 3, fmt.Sprintf("got %q", r3))
	fmt.Println("all tests pass")
}

func main() { runTests() }
